File: python/challenge01/valid_password.py
def is_valid(pwd: str) -> bool:

    state = "start"
    last_digit = "0"
    last_char = "a"

    for c in pwd:
        if "0" <= c <= "9":
            if state == "char":
                return False
            if c < last_digit:
                return False
            last_digit = c
        elif "a" <= c <= "z":
            state = "char"
            if c < last_char:
                return False
            last_char = c
        else:
            return False
    return True


def main():
    valid = 0
    invalid = 0

    with open("log.txt", "r") as file:
        for line in file:
            pwd = line.strip()
            if not pwd:
                continue
            if is_valid(pwd):
                valid += 1
            else:
                invalid += 1
    
    print(f"Submit: {valid}true,{invalid}false")

if __name__ == "__main__":
    main()

File: c/challenge01/valid_password.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

#define MAX_LINE  1024

enum State { START, DIGIT, CHAR };

bool is_valid(const char *pwd) {
    enum State state = START;
    char last_digit = '0';
    char last_char = 'a';

    for (size_t i = 0; pwd[i] != '\0'; i++) {
        char c = pwd[i];

        if (c >= '0' && c <= '9') {
            if (state == CHAR) {
                return false;
            }
            state = DIGIT;
            if (c < last_digit) {
                return false;
            }
            last_digit = c;
        } else if (c >= 'a' && c <= 'z') {
            state = CHAR;
            if (c < last_char) {
                return false;
            }
            last_char = c;
        } else {
            return false;
        }
    }
    return true;
}

int main(void) {
    FILE *fp = fopen("log.txt", "r");
    if (!fp) {
        perror("Failed to open log.txt");
        return EXIT_FAILURE;
    }

    char line[MAX_LINE];
    int valid = 0, invalid = 0;

    while (fgets(line, sizeof(line), fp)) {
        // Remove newline character if present
        line[strcspn(line, "\r\n")] = '\0';
        if (line[0] == '\0') {
            continue;
        }
        if (is_valid(line)) {
            valid++;
        } else {
            invalid++;
        }
    }

    fclose(fp);

    printf("submit %dtrue%dfalse\n", valid, invalid);
    return EXIT_SUCCESS;
}